scripts/opencontext_things.py

```diff
@@ -0,0 +1,155 @@
+import datetime
+import logging
+import click
+import click_config_file
+import asyncio
+import isb_lib.core
+import isb_lib.opencontext_adapter
+import igsn_lib.models
+import concurrent.futures
+import heartrate
+import sqlalchemy
+import sqlalchemy.orm
+import sqlalchemy.exc
+
+LOG_LEVELS = {
+    "DEBUG": logging.DEBUG,
+    "INFO": logging.INFO,
+    "WARNING": logging.WARNING,
+    "WARN": logging.WARNING,
+    "ERROR": logging.ERROR,
+    "FATAL": logging.CRITICAL,
+    "CRITICAL": logging.CRITICAL,
+}
+LOG_DATE_FORMAT = "%Y-%m-%dT%H:%M:%S"
+LOG_FORMAT = "%(asctime)s %(name)s:%(levelname)s: %(message)s"
+BACKLOG_SIZE = 40
+
+
+def get_logger():
+    return logging.getLogger("main")
+
+
+def wrap_load_thing(thing_dict, tc):
+    """Return request information to assist future management"""
+    try:
+        return tc, isb_lib.opencontext_adapter.load_thing(thing_dict)
+    except:
+        pass
+    return tc, None
+
+
+async def _load_open_context_entries(session, max_count, start_from):
+    L = get_logger()
+    records = isb_lib.opencontext_adapter.OpenContextRecordIterator(
+        max_entries=max_count, date_start=start_from, page_size=200
+    )
+
+    num_ids = 0
+    for record in records:
+        L.info("got next id from open context %s", record)
+        num_ids += 1
+        id = record["uri"]
+        try:
+            res = session.query(igsn_lib.models.thing.Thing.id).filter_by(id=id).one()
+            logging.info("Already have %s", id)
+        except sqlalchemy.orm.exc.NoResultFound:
+            thing = isb_lib.opencontext_adapter.load_thing(
+                record, datetime.datetime.now(), records.last_url_str()
+            )
+            try:
+                session.add(thing)
+                session.commit()
+            except sqlalchemy.exc.IntegrityError as e:
+                session.rollback()
+                logging.error("Item already exists: %s", record)
+
+    L.info("total num records %d", num_ids)
+
+    # total_requested = 0
+    # total_completed = 0
+    # more_work = True
+    # num_prepared = BACKLOG_SIZE  # Number of jobs to prepare for execution
+    # with concurrent.futures.ThreadPoolExecutor(
+    #     max_workers=1
+    # ) as executor:
+    #     while more_work:
+    #         while (
+    #             len(futures) < BACKLOG_SIZE
+    #             and total_requested < max_count
+    #             and num_prepared > 0
+    #         ):
+    #             record = next(records)
+    #             print("got next id from open context %s", record)
+
+
+def load_open_context_entries(session, max_count, start_from=None):
+    loop = asyncio.get_event_loop()
+    future = asyncio.ensure_future(
+        _load_open_context_entries(session, max_count, start_from=start_from)
+    )
+    loop.run_until_complete(future)
```

Comment:
I'm not sure there's any point to running this in asyncio as it's currently written, but it was here initially due to it starting life as fork of the GEOME importer.